### PR TITLE
Parallelize wolf energy and virial

### PR DESCRIPTION
**Pull request by antoinewdg**
This PR introduces simple parallelization of energy and virial computation in the Wolf summation method.

### CODE REVIEW

**Comment by antoinewdg** [comment]
## Comparing to master (a1bceef4)
using `--threshold 2, latest commit first`
<details><summary>b2abe064 Parallelize Wolf::virial</summary>
  <p>





```bash

 name                                         a1bceef4.txt ns/iter    b2abe064.txt ns/iter     diff ns/iter   diff % 
 nacl::energy_wolf                            471,927 (+/- 795)       112,303 (+/- 979)            -359,624  -76.20% 
 nacl::virial_wolf                            661,874 (+/- 2613)      143,720 (+/- 5598)           -518,154  -78.29% 
 water::cache_move_all_rigid_molecules_ewald  9,840,395 (+/- 160429)  8,410,492 (+/- 9494)       -1,429,903  -14.53% 
 water::cache_move_all_rigid_molecules_wolf   9,405,560 (+/- 37542)   7,895,312 (+/- 1550908)    -1,510,248  -16.06% 
 water::energy_wolf                           729,992 (+/- 728)       166,272 (+/- 4190)           -563,720  -77.22% 
 water::forces_wolf                           998,767 (+/- 1679)      1,072,697 (+/- 1751)           73,930    7.40% 
 water::virial_wolf                           993,396 (+/- 1307)      212,406 (+/- 18222)          -780,990  -78.62% 



```



</p></details>

<details><summary>6e5c84c1 Parallelize Wolf::energy</summary>
  <p>





```bash

 name                                         a1bceef4.txt ns/iter    6e5c84c1.txt ns/iter     diff ns/iter   diff % 
 nacl::energy_wolf                            471,927 (+/- 795)       109,938 (+/- 2938)           -361,989  -76.70% 
 water::cache_move_all_rigid_molecules_ewald  9,840,395 (+/- 160429)  8,489,054 (+/- 22959)      -1,351,341  -13.73% 
 water::cache_move_all_rigid_molecules_wolf   9,405,560 (+/- 37542)   7,966,530 (+/- 1849291)    -1,439,030  -15.30% 
 water::energy_wolf                           729,992 (+/- 728)       163,003 (+/- 4714)           -566,989  -77.67% 



```



</p></details>

## Individual benchmarks

<details><summary>b2abe064 Parallelize Wolf::virial</summary>
  <p>





```bash


running 4 tests
test argon::cache_move_particle ... bench:     241,311 ns/iter (+/- 1,061)
test argon::energy              ... bench:   1,856,444 ns/iter (+/- 20,072)
test argon::forces              ... bench:   2,354,803 ns/iter (+/- 5,803)
test argon::virial              ... bench:   2,718,919 ns/iter (+/- 3,610)

test result: ok. 0 passed; 0 failed; 0 ignored; 4 measured


running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured


running 8 tests
test nacl::cache_move_particle_ewald ... bench:     408,462 ns/iter (+/- 708)
test nacl::cache_move_particle_wolf  ... bench:     107,101 ns/iter (+/- 20,390)
test nacl::energy_ewald              ... bench:     617,827 ns/iter (+/- 561)
test nacl::energy_wolf               ... bench:     112,303 ns/iter (+/- 979)
test nacl::forces_ewald              ... bench:  17,207,822 ns/iter (+/- 41,245)
test nacl::forces_wolf               ... bench:     662,433 ns/iter (+/- 649)
test nacl::virial_ewald              ... bench:  48,059,408 ns/iter (+/- 23,181)
test nacl::virial_wolf               ... bench:     143,720 ns/iter (+/- 5,598)

test result: ok. 0 passed; 0 failed; 0 ignored; 8 measured


running 5 tests
test propane::cache_move_all_rigid_molecules ... bench:   1,027,511 ns/iter (+/- 12,972)
test propane::cache_move_particles           ... bench:     280,059 ns/iter (+/- 11,146)
test propane::energy                         ... bench:   1,114,225 ns/iter (+/- 1,996)
test propane::forces                         ... bench:   1,379,765 ns/iter (+/- 2,383)
test propane::virial                         ... bench:   1,441,683 ns/iter (+/- 2,805)

test result: ok. 0 passed; 0 failed; 0 ignored; 5 measured


running 10 tests
test water::cache_move_all_rigid_molecules_ewald ... bench:   8,410,492 ns/iter (+/- 9,494)
test water::cache_move_all_rigid_molecules_wolf  ... bench:   7,895,312 ns/iter (+/- 1,550,908)
test water::cache_move_particles_ewald           ... bench:     858,383 ns/iter (+/- 977)
test water::cache_move_particles_wolf            ... bench:     484,795 ns/iter (+/- 56,672)
test water::energy_ewald                         ... bench:   1,019,537 ns/iter (+/- 1,417)
test water::energy_wolf                          ... bench:     166,272 ns/iter (+/- 4,190)
test water::forces_ewald                         ... bench:  24,851,752 ns/iter (+/- 31,893)
test water::forces_wolf                          ... bench:   1,072,697 ns/iter (+/- 1,751)
test water::virial_ewald                         ... bench:  69,412,618 ns/iter (+/- 14,700)
test water::virial_wolf                          ... bench:     212,406 ns/iter (+/- 18,222)

test result: ok. 0 passed; 0 failed; 0 ignored; 10 measured




```



</p></details>

<details><summary>6e5c84c1 Parallelize Wolf::energy</summary>
  <p>





```bash


running 4 tests
test argon::cache_move_particle ... bench:     241,877 ns/iter (+/- 1,148)
test argon::energy              ... bench:   1,842,226 ns/iter (+/- 2,041)
test argon::forces              ... bench:   2,355,309 ns/iter (+/- 2,008)
test argon::virial              ... bench:   2,703,837 ns/iter (+/- 2,122)

test result: ok. 0 passed; 0 failed; 0 ignored; 4 measured


running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured


running 8 tests
test nacl::cache_move_particle_ewald ... bench:     410,504 ns/iter (+/- 2,669)
test nacl::cache_move_particle_wolf  ... bench:     106,062 ns/iter (+/- 7,002)
test nacl::energy_ewald              ... bench:     622,801 ns/iter (+/- 2,312)
test nacl::energy_wolf               ... bench:     109,938 ns/iter (+/- 2,938)
test nacl::forces_ewald              ... bench:  17,241,779 ns/iter (+/- 29,406)
test nacl::forces_wolf               ... bench:     664,724 ns/iter (+/- 1,647)
test nacl::virial_ewald              ... bench:  48,381,327 ns/iter (+/- 71,926)
test nacl::virial_wolf               ... bench:     660,850 ns/iter (+/- 1,221)

test result: ok. 0 passed; 0 failed; 0 ignored; 8 measured


running 5 tests
test propane::cache_move_all_rigid_molecules ... bench:   1,037,861 ns/iter (+/- 13,536)
test propane::cache_move_particles           ... bench:     281,062 ns/iter (+/- 10,625)
test propane::energy                         ... bench:   1,119,935 ns/iter (+/- 1,474)
test propane::forces                         ... bench:   1,384,211 ns/iter (+/- 12,890)
test propane::virial                         ... bench:   1,442,179 ns/iter (+/- 1,080)

test result: ok. 0 passed; 0 failed; 0 ignored; 5 measured


running 10 tests
test water::cache_move_all_rigid_molecules_ewald ... bench:   8,489,054 ns/iter (+/- 22,959)
test water::cache_move_all_rigid_molecules_wolf  ... bench:   7,966,530 ns/iter (+/- 1,849,291)
test water::cache_move_particles_ewald           ... bench:     861,948 ns/iter (+/- 1,372)
test water::cache_move_particles_wolf            ... bench:     488,403 ns/iter (+/- 266,623)
test water::energy_ewald                         ... bench:   1,016,773 ns/iter (+/- 1,884)
test water::energy_wolf                          ... bench:     163,003 ns/iter (+/- 4,714)
test water::forces_ewald                         ... bench:  25,027,911 ns/iter (+/- 69,418)
test water::forces_wolf                          ... bench:     999,045 ns/iter (+/- 1,904)
test water::virial_ewald                         ... bench:  69,871,490 ns/iter (+/- 275,088)
test water::virial_wolf                          ... bench:     997,268 ns/iter (+/- 35,331)

test result: ok. 0 passed; 0 failed; 0 ignored; 10 measured




```



</p></details>

<details><summary>a1bceef4 Update dependencies for parallelization</summary>
  <p>





```bash


running 4 tests
test argon::cache_move_particle ... bench:     241,481 ns/iter (+/- 1,930)
test argon::energy              ... bench:   1,849,778 ns/iter (+/- 2,298)
test argon::forces              ... bench:   2,353,579 ns/iter (+/- 2,235)
test argon::virial              ... bench:   2,710,618 ns/iter (+/- 2,279)

test result: ok. 0 passed; 0 failed; 0 ignored; 4 measured


running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured


running 8 tests
test nacl::cache_move_particle_ewald ... bench:     409,231 ns/iter (+/- 628)
test nacl::cache_move_particle_wolf  ... bench:     106,201 ns/iter (+/- 109)
test nacl::energy_ewald              ... bench:     619,715 ns/iter (+/- 329)
test nacl::energy_wolf               ... bench:     471,927 ns/iter (+/- 795)
test nacl::forces_ewald              ... bench:  17,257,931 ns/iter (+/- 31,278)
test nacl::forces_wolf               ... bench:     665,139 ns/iter (+/- 895)
test nacl::virial_ewald              ... bench:  48,371,348 ns/iter (+/- 13,781)
test nacl::virial_wolf               ... bench:     661,874 ns/iter (+/- 2,613)

test result: ok. 0 passed; 0 failed; 0 ignored; 8 measured


running 5 tests
test propane::cache_move_all_rigid_molecules ... bench:   1,029,671 ns/iter (+/- 16,776)
test propane::cache_move_particles           ... bench:     281,081 ns/iter (+/- 10,922)
test propane::energy                         ... bench:   1,116,074 ns/iter (+/- 40,962)
test propane::forces                         ... bench:   1,380,016 ns/iter (+/- 72,546)
test propane::virial                         ... bench:   1,440,185 ns/iter (+/- 1,479)

test result: ok. 0 passed; 0 failed; 0 ignored; 5 measured


running 10 tests
test water::cache_move_all_rigid_molecules_ewald ... bench:   9,840,395 ns/iter (+/- 160,429)
test water::cache_move_all_rigid_molecules_wolf  ... bench:   9,405,560 ns/iter (+/- 37,542)
test water::cache_move_particles_ewald           ... bench:     856,290 ns/iter (+/- 1,988)
test water::cache_move_particles_wolf            ... bench:     482,736 ns/iter (+/- 2,701)
test water::energy_ewald                         ... bench:   1,018,571 ns/iter (+/- 2,638)
test water::energy_wolf                          ... bench:     729,992 ns/iter (+/- 728)
test water::forces_ewald                         ... bench:  24,920,899 ns/iter (+/- 32,880)
test water::forces_wolf                          ... bench:     998,767 ns/iter (+/- 1,679)
test water::virial_ewald                         ... bench:  69,749,505 ns/iter (+/- 44,531)
test water::virial_wolf                          ... bench:     993,396 ns/iter (+/- 1,307)

test result: ok. 0 passed; 0 failed; 0 ignored; 10 measured




```



</p></details>

<br>**CPU**: Intel(R) Core(TM) i7-5700HQ CPU @ 2.70GHz

**Comment by Luthaf** [comment]
## Comparing to master (a1bceef4)
using `--threshold 2, latest commit first`
<details><summary>b2abe064 Parallelize Wolf::virial</summary>
  <p>





```bash

 name                                        a1bceef4.txt ns/iter    b2abe064.txt ns/iter     diff ns/iter   diff % 
 nacl::energy_wolf                           960,241 (+/- 549)       98,812 (+/- 1437)            -861,429  -89.71% 
 nacl::virial_wolf                           1,174,982 (+/- 651)     114,710 (+/- 1994)         -1,060,272  -90.24% 
 water::cache_move_all_rigid_molecules_wolf  12,519,146 (+/- 99904)  11,687,572 (+/- 552750)      -831,574   -6.64% 
 water::energy_wolf                          1,426,281 (+/- 987)     132,648 (+/- 7297)         -1,293,633  -90.70% 
 water::virial_wolf                          1,754,964 (+/- 816)     163,892 (+/- 14033)        -1,591,072  -90.66% 



```



</p></details>

<details><summary>6e5c84c1 Parallelize Wolf::energy</summary>
  <p>





```bash

 name                                        a1bceef4.txt ns/iter    6e5c84c1.txt ns/iter     diff ns/iter   diff % 
 nacl::energy_wolf                           960,241 (+/- 549)       101,723 (+/- 2695)           -858,518  -89.41% 
 water::cache_move_all_rigid_molecules_wolf  12,519,146 (+/- 99904)  11,930,504 (+/- 972226)      -588,642   -4.70% 
 water::cache_move_particles_wolf            749,011 (+/- 7764)      784,058 (+/- 158716)           35,047    4.68% 
 water::energy_wolf                          1,426,281 (+/- 987)     131,290 (+/- 7109)         -1,294,991  -90.79% 



```



</p></details>

## Individual benchmarks

<details><summary>b2abe064 Parallelize Wolf::virial</summary>
  <p>





```bash


running 4 tests
test argon::cache_move_particle ... bench:     326,548 ns/iter (+/- 3,636)
test argon::energy              ... bench:   3,979,673 ns/iter (+/- 24,271)
test argon::forces              ... bench:   4,502,489 ns/iter (+/- 5,805)
test argon::virial              ... bench:   5,363,281 ns/iter (+/- 5,233)

test result: ok. 0 passed; 0 failed; 0 ignored; 4 measured


running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured


running 8 tests
test nacl::cache_move_particle_ewald ... bench:     654,363 ns/iter (+/- 1,012)
test nacl::cache_move_particle_wolf  ... bench:     166,359 ns/iter (+/- 2,796)
test nacl::energy_ewald              ... bench:   1,157,217 ns/iter (+/- 1,934)
test nacl::energy_wolf               ... bench:      98,812 ns/iter (+/- 1,437)
test nacl::forces_ewald              ... bench:  25,329,671 ns/iter (+/- 91,705)
test nacl::forces_wolf               ... bench:   1,136,546 ns/iter (+/- 775)
test nacl::virial_ewald              ... bench: 106,997,190 ns/iter (+/- 1,288,941)
test nacl::virial_wolf               ... bench:     114,710 ns/iter (+/- 1,994)

test result: ok. 0 passed; 0 failed; 0 ignored; 8 measured


running 5 tests
test propane::cache_move_all_rigid_molecules ... bench:   1,993,360 ns/iter (+/- 4,660)
test propane::cache_move_particles           ... bench:     513,131 ns/iter (+/- 6,161)
test propane::energy                         ... bench:   2,194,563 ns/iter (+/- 20,403)
test propane::forces                         ... bench:   2,565,543 ns/iter (+/- 15,541)
test propane::virial                         ... bench:   2,733,750 ns/iter (+/- 2,682)

test result: ok. 0 passed; 0 failed; 0 ignored; 5 measured


running 10 tests
test water::cache_move_all_rigid_molecules_ewald ... bench:  13,042,694 ns/iter (+/- 32,339)
test water::cache_move_all_rigid_molecules_wolf  ... bench:  11,687,572 ns/iter (+/- 552,750)
test water::cache_move_particles_ewald           ... bench:   1,362,510 ns/iter (+/- 8,016)
test water::cache_move_particles_wolf            ... bench:     760,013 ns/iter (+/- 57,921)
test water::energy_ewald                         ... bench:   1,876,289 ns/iter (+/- 1,978)
test water::energy_wolf                          ... bench:     132,648 ns/iter (+/- 7,297)
test water::forces_ewald                         ... bench:  37,103,394 ns/iter (+/- 104,036)
test water::forces_wolf                          ... bench:   1,729,404 ns/iter (+/- 1,503)
test water::virial_ewald                         ... bench: 158,452,012 ns/iter (+/- 42,382)
test water::virial_wolf                          ... bench:     163,892 ns/iter (+/- 14,033)

test result: ok. 0 passed; 0 failed; 0 ignored; 10 measured




```



</p></details>

<details><summary>6e5c84c1 Parallelize Wolf::energy</summary>
  <p>





```bash


running 4 tests
test argon::cache_move_particle ... bench:     329,154 ns/iter (+/- 2,854)
test argon::energy              ... bench:   4,023,117 ns/iter (+/- 21,803)
test argon::forces              ... bench:   4,503,138 ns/iter (+/- 5,003)
test argon::virial              ... bench:   5,384,481 ns/iter (+/- 6,602)

test result: ok. 0 passed; 0 failed; 0 ignored; 4 measured


running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured


running 8 tests
test nacl::cache_move_particle_ewald ... bench:     654,297 ns/iter (+/- 1,027)
test nacl::cache_move_particle_wolf  ... bench:     168,342 ns/iter (+/- 6,681)
test nacl::energy_ewald              ... bench:   1,156,220 ns/iter (+/- 1,465)
test nacl::energy_wolf               ... bench:     101,723 ns/iter (+/- 2,695)
test nacl::forces_ewald              ... bench:  25,347,807 ns/iter (+/- 99,097)
test nacl::forces_wolf               ... bench:   1,131,349 ns/iter (+/- 630)
test nacl::virial_ewald              ... bench: 107,309,053 ns/iter (+/- 1,013,440)
test nacl::virial_wolf               ... bench:   1,172,859 ns/iter (+/- 905)

test result: ok. 0 passed; 0 failed; 0 ignored; 8 measured


running 5 tests
test propane::cache_move_all_rigid_molecules ... bench:   1,986,320 ns/iter (+/- 5,838)
test propane::cache_move_particles           ... bench:     514,372 ns/iter (+/- 3,209)
test propane::energy                         ... bench:   2,200,101 ns/iter (+/- 22,355)
test propane::forces                         ... bench:   2,560,355 ns/iter (+/- 15,873)
test propane::virial                         ... bench:   2,735,546 ns/iter (+/- 2,125)

test result: ok. 0 passed; 0 failed; 0 ignored; 5 measured


running 10 tests
test water::cache_move_all_rigid_molecules_ewald ... bench:  13,051,019 ns/iter (+/- 15,588)
test water::cache_move_all_rigid_molecules_wolf  ... bench:  11,930,504 ns/iter (+/- 972,226)
test water::cache_move_particles_ewald           ... bench:   1,364,127 ns/iter (+/- 4,242)
test water::cache_move_particles_wolf            ... bench:     784,058 ns/iter (+/- 158,716)
test water::energy_ewald                         ... bench:   1,877,779 ns/iter (+/- 1,705)
test water::energy_wolf                          ... bench:     131,290 ns/iter (+/- 7,109)
test water::forces_ewald                         ... bench:  37,115,432 ns/iter (+/- 107,859)
test water::forces_wolf                          ... bench:   1,720,995 ns/iter (+/- 1,238)
test water::virial_ewald                         ... bench: 157,156,763 ns/iter (+/- 52,699)
test water::virial_wolf                          ... bench:   1,755,676 ns/iter (+/- 906)

test result: ok. 0 passed; 0 failed; 0 ignored; 10 measured




```



</p></details>

<details><summary>a1bceef4 Update dependencies for parallelization</summary>
  <p>





```bash


running 4 tests
test argon::cache_move_particle ... bench:     328,205 ns/iter (+/- 3,400)
test argon::energy              ... bench:   3,998,307 ns/iter (+/- 10,104)
test argon::forces              ... bench:   4,505,166 ns/iter (+/- 4,904)
test argon::virial              ... bench:   5,390,215 ns/iter (+/- 5,785)

test result: ok. 0 passed; 0 failed; 0 ignored; 4 measured


running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured


running 8 tests
test nacl::cache_move_particle_ewald ... bench:     655,773 ns/iter (+/- 1,147)
test nacl::cache_move_particle_wolf  ... bench:     167,119 ns/iter (+/- 256)
test nacl::energy_ewald              ... bench:   1,155,038 ns/iter (+/- 1,693)
test nacl::energy_wolf               ... bench:     960,241 ns/iter (+/- 549)
test nacl::forces_ewald              ... bench:  25,313,724 ns/iter (+/- 99,688)
test nacl::forces_wolf               ... bench:   1,140,772 ns/iter (+/- 733)
test nacl::virial_ewald              ... bench: 107,264,965 ns/iter (+/- 1,324,331)
test nacl::virial_wolf               ... bench:   1,174,982 ns/iter (+/- 651)

test result: ok. 0 passed; 0 failed; 0 ignored; 8 measured


running 5 tests
test propane::cache_move_all_rigid_molecules ... bench:   2,009,139 ns/iter (+/- 5,475)
test propane::cache_move_particles           ... bench:     514,768 ns/iter (+/- 2,836)
test propane::energy                         ... bench:   2,196,019 ns/iter (+/- 21,486)
test propane::forces                         ... bench:   2,559,019 ns/iter (+/- 16,510)
test propane::virial                         ... bench:   2,738,716 ns/iter (+/- 1,581)

test result: ok. 0 passed; 0 failed; 0 ignored; 5 measured


running 10 tests
test water::cache_move_all_rigid_molecules_ewald ... bench:  13,101,466 ns/iter (+/- 20,579)
test water::cache_move_all_rigid_molecules_wolf  ... bench:  12,519,146 ns/iter (+/- 99,904)
test water::cache_move_particles_ewald           ... bench:   1,370,352 ns/iter (+/- 10,492)
test water::cache_move_particles_wolf            ... bench:     749,011 ns/iter (+/- 7,764)
test water::energy_ewald                         ... bench:   1,877,600 ns/iter (+/- 1,361)
test water::energy_wolf                          ... bench:   1,426,281 ns/iter (+/- 987)
test water::forces_ewald                         ... bench:  37,097,310 ns/iter (+/- 86,902)
test water::forces_wolf                          ... bench:   1,722,913 ns/iter (+/- 1,285)
test water::virial_ewald                         ... bench: 156,867,801 ns/iter (+/- 43,149)
test water::virial_wolf                          ... bench:   1,754,964 ns/iter (+/- 816)

test result: ok. 0 passed; 0 failed; 0 ignored; 10 measured




```



</p></details>

<br>**CPU**: Intel(R) Xeon(R) CPU E5-2660 0 @ 2.20GHz

File: src/core/src/energy/global/wolf.rs [hunk 1]
```diff
@@ -172,10 +174,13 @@ impl GlobalPotential for Wolf {
 
     fn energy(&self, system: &System) -> f64 {
         let natoms = system.size();
-        let mut res = 0.0;
-        for i in 0..natoms {
+
+        (0..natoms).into_par_iter().map(|i|{
```
Comment (Luthaf) [inline src/core/src/energy/global/wolf.rs]: As it seems to be a common pattern in the code, how about having a function: `range_par(Range<usize>)` to use in `range_par(0..natoms).map(|i| ...` ? Or even `range_par(natoms).map(|i| ...`?

File: src/core/src/energy/global/wolf.rs [hunk 2]
```diff
@@ -229,10 +236,11 @@ impl GlobalPotential for Wolf {
...
                 let rij = system.nearest_image(i, j);
                 let force = self.force_pair(info, qi, qj, rij);
-                res += force.tensorial(&rij);
+                virial += force.tensorial(&rij);
             }
-        }
-        return res;
+
+            virial
+        }).reduce(|| Matrix3::zero(), |a, b| a + b)
```
Comment (Luthaf) [inline src/core/src/energy/global/wolf.rs]: Why not use `sum` here?

File: src/core/src/energy/global/wolf.rs [hunk 3]
```diff
@@ -216,10 +220,13 @@ impl GlobalPotential for Wolf {
...
         let natoms = system.size();
-        let mut res = Matrix3::zero();
-        for i in 0..natoms {
+
+        (0..natoms).into_par_iter().map(|i| {
+
             let qi = system[i].charge;
-            if qi == 0.0 {continue}
+            if qi == 0.0 { return Matrix3::zero(); }
+            let mut virial = Matrix3::zero();
```
Comment (Luthaf) [inline src/core/src/energy/global/wolf.rs]: Should we call all local variables `local_***`? I don't care, as long as it is consistent in all the code (Ewald uses local_virial)

Comment (antoinewdg) [inline src/core/src/energy/global/wolf.rs]: Sure

File: src/core/src/energy/global/wolf.rs (hunk 1 above)
Comment (antoinewdg) [inline src/core/src/energy/global/wolf.rs]: That indeed happens often. Maybe in the same idea as `Zip` in `ndarray_parallel`, have a `par_map` shortcut that does this.
We would have `(0..nations).par_map(|i| ...)`.

Comment (Luthaf) [inline src/core/src/energy/global/wolf.rs]: Yes, this looks nice =)

File: src/core/src/energy/global/wolf.rs [hunk 4]
```diff
@@ -4,10 +4,13 @@
 use special::Error;
 use std::f64::consts::PI;
 
+use rayon::prelude::*;
```
Comment (Luthaf) [inline src/core/src/energy/global/wolf.rs]: Do we need this and `ParallelShortcuts`?

**Comment by Luthaf** [comment]
The coverage drop is strange. It looks like the Ewald code is no longer tested, or codecov does not pick it up.

File: src/core/src/energy/global/wolf.rs (hunk 4 above)
Comment (antoinewdg) [inline src/core/src/energy/global/wolf.rs]: If we go with our own utils::parallel we could have our own prelude here

**Comment by antoinewdg** [comment]
I don't get it either, it seems comments are counted in the coverage :(.
I'll restart the buid.

**Comment by Luthaf** [comment]
It could be https://users.rust-lang.org/t/kcov-reports-less-coverage-after-rust-1-17-released/10675. It might go away by rebasing on master.

**Comment by antoinewdg** [comment]
I rebased, if this passes it's ready to merge on my side.

**Pull request merged**